keps/sig-cluster-lifecycle/draft-20180316-kubeadm-phases-beta.md

```diff
@@ -0,0 +1,243 @@
+commands to beta (and GA) and, at the same time, to enable a significant reduction
+of the effort for maintaining this feature.
+
+### Goals
+
+* To define the approach for graduating to beta (GA) the `kubeadm phase` user
+  interface.
+* To define a new approach for implementing the `kubeadm phase` user interface
+  that significantly reduces the effort for maintaining this feature up to date
+  with the continuous changes of the init workflow.
+* To provide a centralized point for controlling the proliferation of flags in
+  kubeadm.
+* To avoid inconsistencies between init and phases (both at UX and implementation
+  logic level)
+
+### Non-goals
+
+* This proposal doesn't include any changes/improvement to the init workflow
+  internal logic.
+* Nothing in this proposal should prevent practices that exist today.
+
+### Challenges and Open Questions
+
+* How to reduce the effort for maintaining `kubeadm phase` feature up to date
+  with the continuous changes of the init workflow?
+* How to provide an intuitive user interface for both phases and sub-phases
+  (e.g  `kubeadm alpha phase certs` and `kubeadm alpha phase certs apiserver`)?
+* Should we preserve all the second/third level nested phases implemented today
+  (e.g. certs CA, certs api-server, certs sa etc.), or should we allow only second
+  level phases (e.g certs), or should we maintain & simplify second/third level phases?
+* Should we preserve or deprecate the (few) phases implemented today that are not
+  part of the `kubeadm init` workflow? If preserve, in which form? as a new first
+  level commands?
+* While `kubeadm init` executes some phases conditionally (e.g. self-hosting only
+  if the corresponding feature-gate enable), the current implementation of phases
+  executes them unconditionally. Should we preserve this behavior?
+* How to preserve the possibility to automatically generate kubeadm documentation?
+* How to address/enable the possibility to generate machine readable output from
+  kubeadm commands?
+
+## Proposal
+
+### User Stories
+
+#### Invoke a specific task/phase of the kubeadm init workflow
+
+As a kubernetes administrator, I want to run only a specific task/phase of the
+`kubeadm init` workflow.
+
+### Implementation Details
+
+In order to achieve the goals of this proposal the approach suggested is to customize
+the `kubeadm init` command transforming it's usage from:
+
+```bash
+Usage:
+  kubeadm init [flags]
+```
+
+to
+
+```bash
+Usage:
+  kubeadm init [phases] [flags]
+```
+
+[phases] will be a new, optional, argument that allows user to select which part/phase
+of the `kubeadm init`  logic to execute.
+
+As a consequence:
+
+* To invoke the beta (GA) version of  `kubeadm alpha phase certs` the user will be required
+  to execute:
+
+```bash
+kubeadm init certs [flags]
+```
+
+* To invoke the beta (GA) version of  `kubeadm alpha phase certs apiserver` the user
+  will be required to execute:
+
+```bash
+kubeadm init certs/apiserver [flags]
```

Comment:
is there prior art for the `phase/subphase` syntax anywhere else in kubeadm / k8s in general?

Reply:
Strictly speaking of phase/subphase of a command, there isn't nothing I'm aware of.
If we consider the path generic pattern for expressing a hierarchy of things, it is already present in K8s e.g. in `kubectl scale deployment/nginx --replicas=3`

Reply:
How about more explicit naming, e.g. something like `allCerts`, `apiserverCerts` etc? I find that characters with special meanings in command-line arguments can be very misleading, someone might think phases map to files, or even consider that it's similar to `deployment/nginx`...

Reply:
I think it's a tricky balance to strike. Having all those separate commands is going to make discovery / organising. Having the explicit hierarchy makes sense to me. I was hoping for something simpler like `kubeadm init certs apiserver`, but that obviates being able to run multiple commands. I like the implementation as-is, the prior art that exists is good enough for me.

Reply:
I wasn't clear enough, I was assuming implying my other suggestion to introduce an explicit `--only` flag, so the usage would be something like `--only= allCerts` or `--only=apiserverCerts`. To discover phases `--only=help` could be an option, or otherwise `--help` should be sufficient actually.

Reply:
> I like the implementation as-is, the prior art that exists is good enough for me.

It seems mostly okay to me too (except a few odd naming decisions).

However, if we want to move this functionality under `kubeadm init`, I'd strongly insist on having an explicit flag and an unambiguous naming scheme.

Reply:
@errordeveloper message received strong and clear. I start tinkering around this